_posts/2017-11-23-totw-126.md

```diff
@@ -0,0 +1,117 @@
+    value in minimizing calls to `new`. More on this in \#5.
+
+4.  If all allocations are handled via `absl::make_unique()` or factory calls,
+    that leaves `absl::WrapUnique()` for the implementation of those factory
+    calls, for code interacting with legacy methods that don't rely on
+    `std::unique_ptr` for ownership transfer, and for rare cases that need to
+    dynamically allocate with aggregate initialization (`absl::WrapUnique(new
+    MyStruct{3.141, "pi"})`). In code review it's easy to spot the
+    `absl::WrapUnique` calls and evaluate "does that expression look like an
+    ownership transfer?" Usually it's obvious (for example, it's some factory
+    function). When it's not obvious, we need to check the function to be sure
+    that it's actually a raw-pointer ownership transfer.
+
+5.  If we instead rely mostly on the constructors of `std::unique_ptr`, we see
+    calls like: \
+    `std::unique_ptr<T> foo(Blah());` \
+    `std::unique_ptr<T> bar(new T());` \
+    It takes only a moment's inspection to see that the latter is safe (no leak,
+    no double-delete). The former? It depends: if `Blah()` is returning a
+    `std::unique_ptr`, it's fine, though in that case it would be more obviously
+    safe if written as \
+    `std::unique_ptr<T> foo = Blah();` \
+    If `Blah()` is returning an ownership-transferred raw pointer, that's also
+    fine. If `Blah()` is returning just some random pointer (no transfer), then
+    there's a problem. Reliance on `absl::make_unique()` and `absl::WrapUnique()`
+    (avoiding the constructors) provides an additional visual clue for the
+    places where we have to worry (calls to `absl::WrapUnique()`, and only those).
+
+## How Should We Choose Which to Use?
+
+1.  By default, use `absl::make_unique()` (or `std::make_shared()` for the rare
+    cases where shared ownership is appropriate) for dynamic allocation. For
+    example, instead of: `std::unique_ptr<T> bar(new T());` write `auto bar
+    = absl::make_unique<T>();` and instead of `bar.reset(new T());` write
+    `bar = absl::make_unique<T>();`
+
+2.  In a factory function that uses a non-public constructor, return a
+    `std::unique_ptr<T>` and use `absl::WrapUnique(new T(...))` in the
+    implementation.
+
+3.  When dynamically allocating an object that requires brace initialization
+    (typically a struct, an array, or a container), use `absl::WrapUnique(new
+    T{...})`.
+
+4.  When calling a legacy API that accepts ownership via a `T*`, either allocate
+    the object in advance with `absl::make_unique` and call `ptr.release()` in
+    the call, or use `new` directly in the function argument.
+
+5.  When calling a legacy API that returns ownership via a `T*`, immediately
+    construct a smart pointer with `WrapUnique` (unless you're immediately
+    passing the pointer to another legacy API that accepts ownership via a
+    `T*`).
+
+## Summary
+Prefer `absl::make_unique()` over `absl::WrapUnique()`, and prefer
```

Comment:
Add blank line between heading and sentence.

Reply:
Done